### ISSUE TITLE
JMSToolBox.exe   is signed with expired certificate and as such unable to launch due to policy

### ISSUE THREAD

**Issue by guidovossen** [jmstoolbox/jmstoolbox#173]
## Specifications
- JMSToolBox version: jmstoolbox-6.6.0-win32.win32.x86_64
- OS (Name, version, 32/64bits): win
- Q provider (kind, version) : n/a

## Expected Behavior
that i can open it,  in my company we cannot execute unsigned executables. the jmstoolbox.exe is signed with a cert that is having a valid to date in the past. 

‎Wednesday, ‎May ‎22, ‎2024 1:59:59 AM

![image](https://github.com/jmstoolbox/jmstoolbox/assets/48794999/a0f5876a-717f-4261-9be8-c9ac9dab4fa4)

## Steps to Reproduce the Problem

check certicate by going to file -> properties -> digital signature.

## Logs
Eventually, attach the log files which are `"<user>\.jtb\JMSToolBox\jmstoolbox.log"` and `"<user>\.jtb\.metadata\.log"`    
They can be exported directly from JMStoolBox: Menu "Help" > "Export JMSToolBox log files..."

**Comment by titou10titou10** [comment]
That's weird as the JMSToolBox/eclipse.exe is directly build from eclipse and renamed by eclipse-tycho...
I'll have to check where the .exe is coming from and maybe open a ticket on eclipse...

**Comment by guidovossen** [comment]
thanks Denis ! 
would  love to use your tool!

**Comment by guidovossen** [comment]
@titou10titou10   Hi Denis, appreciate your effort. Do you know if eclipse.org could build a new  new executable that would be signed with a new cert?

**Comment by titou10titou10** [comment]
I do not know. eclipse.exe downloaded from eclipse.org is coprrect, howver the package used by tycho that handle the executable bundled with rcp applications is not. this affect other apps built on eclipse RCP,  like DBeaver for example
I'm not sure where to look out

**Comment by titou10titou10** [comment]
I found this: https://github.com/eclipse-simrel/simrel.build/issues/363

**Comment by mbarbero** [comment]
The digital signature of an artifact is valid if the signing time falls within the certificate's validity period. This holds true even after the certificate expires. In the provided screenshot, this condition is met, and Windows reports the signature as valid accordingly.

**Comment by jonahgraham** [comment]
The issue you are having is that you are modifying launcher.exe (Aka eclipse.exe / JMSToolBox.exe) as part of your build, but if you do that you need to resign the JMSToolBox.exe. The modification is:

https://github.com/jmstoolbox/jmstoolbox/blob/b9cd07ca350049c2d1bc7e92cf2ff04ef8095a3a/org.titou10.jtb.product/org.titou10.jtb.product#L27-L30

where you add your own icon. JMSToolBoxc.exe doesn't have modified icon (which is why the screenshot from OP shows valid signature as it was for JMSToolBoxc.exe, not JMSToolBox.exe)

If you look at any older version of JMXToolBox.exe (e.g. the one in 6.3.0) it will also report as invalid. It just seems that no one has reported this problem until now.

Your options are:

1. Remove icon:

```diff
$ git diff
diff --git a/org.titou10.jtb.product/org.titou10.jtb.product b/org.titou10.jtb.product/org.titou10.jtb.product
index c723742..4b80f11 100644
--- a/org.titou10.jtb.product/org.titou10.jtb.product
+++ b/org.titou10.jtb.product/org.titou10.jtb.product
@@ -24,10 +24,6 @@
       location="org.titou10.jtb.core" />
    <launcher name="JMSToolBox">
       <linux icon="icons/jms-icon.xpm"/>
-      <win useIco="true">
-         <ico path="icons/denis_jms-icon.ico"/>
-         <bmp/>
-      </win>
    </launcher>
 
    <vm>
```

2. set up Windows signing infrastructure and get a certificate so you can continue to modify the exe and sign it.
3. Ask affected people to use the JMXToolBoxc.exe (perhaps as a workaround until something else is done)

I am sure others have had similar issues in the past, I recommend checking out [discussions](https://github.com/eclipse-tycho/tycho/discussions) in eclipse tycho.

**Comment by titou10titou10** [comment]
@jonahgraham I just arrived at the same conclusion in parallel of you lol

**Comment by titou10titou10** [comment]
On top JMSToolBox.xex with an icon, on the bottom without
![ddd](https://github.com/jmstoolbox/jmstoolbox/assets/15241034/7afc04f9-9bfa-47e7-a79b-f4348a948eda)

**Comment by titou10titou10** [comment]
@guidovossen is that OK for you to use`JMSToolBoxc.exe`instead of `JMSToolBox.exe`i

**Comment by guidovossen** [comment]
@titou10titou10  really appreciate the effort you put into it! yes that is correct. the certificate itself is valid, but the root certificate that signed it is now expired. At the moment of building the exe 7th may 2024, the root CA cert was not yet expired, but that expired 24th may 2024.  so that is why you would be able to create and sign the JMSToolboxc.exe. 

we have some intrusive local agents in our company and not only check the validity of the certificate that signed the exe but also the validity of the root certificate. It is that that is now expired.   Which makes it impossible to use the tool. I would be suprised if I'm the only one with such a strict company policy.

what if you make a small cosmetic change that triggers a new build? Assuming that the builders replaced their expired root CA cert, it would be build and signed by means of a  (hopefully) refreshed root CA cert.

**Comment by jonahgraham** [comment]
> what if you make a small cosmetic change that triggers a new build?

The issue faced by jmstoolbox is that it doesn't do any signing of the exes at all. It relies on Eclipse providing it signed content. Currently Eclipse doesn't plan to release a new version that would have an updated signature until September, and that release will only have a new signature if the content of launcher.exe (aka eclipse.exe/jmstoolbox.exe) has actually changed.

To me the security policy at your company seems incompatible with how Eclipse treats signing. Are you able to provide more information? e.g. what the software is that is providing the error, what the error is (screenshot?) etc?

I am the chair of the Eclipse Planning Council and the builder of the Eclipse Packaging Project and I am interested in bringing this issue back to Eclipse to see what we can do about it, but I will need more details to understand why it is failing for you (@guidovossen) first.

---

In the meantime I do have a couple of additional workarounds to offer. 

### download latest eclipse.exe

launcher.exe (aka eclipse.exe/jmstoolbox.exe) does not change often and is normally compatible across versions of Eclipse IDE. It has recently been recompiled, but not yet released. The recompiled version can be downloaded, renamed to jmstoolbox.exe and used instead of the one that jmstoolbox.exe ships. Download it from the official [github repo here](https://github.com/eclipse-equinox/equinox.binaries/blob/04b19a00a52e749840201abc25b9ea81726b6eb6/org.eclipse.equinox.executable/bin/win32/win32/x86_64/eclipse.exe) or for console version [here](https://github.com/eclipse-equinox/equinox.binaries/blob/04b19a00a52e749840201abc25b9ea81726b6eb6/org.eclipse.equinox.executable/bin/win32/win32/x86_64/eclipsec.exe).

### bypass eclipse.exe and launch Java directly

The other option is to bypass eclipse.exe entirely and launch java directly. Something like this should work *you can't use java provided by [Eclipse Temurin](https://adoptium.net/temurin/releases/) either as it is also signed with the same certificate, you'll need to look elsewhere for that*:

`javaw --add-modules=ALL-SYSTEM -Djava.security.manager=allow -jar plugins/org.eclipse.equinox.launcher_1.6.600.v20231106-1826.jar`

You may need to add additional VM args, have a look in jmstoolbox.ini for what they are.

**Comment by guidovossen** [comment]
@jonahgraham  
we use Beyondtrust  and while it is whitelisted by my org, it is still disallowing because of expired cert: 

![image](https://github.com/jmstoolbox/jmstoolbox/assets/48794999/e4632363-5d07-4d48-8276-0491905d891e)


now, i never tried the the JMSToolboxc.exe and that seemed to run :) 

![image](https://github.com/jmstoolbox/jmstoolbox/assets/48794999/d8285b70-3627-429d-9296-19e357f3e560)


so , eh it looks like I don't have a problem anymore. I must admit that i do know the difference between JMSToolbox.exe and JMSToolbox**c**.exe

**Comment by titou10titou10** [comment]
> Currently Eclipse doesn't plan to release a new version that would have an updated signature until September, and that release will only have a new signature if the content of launcher.exe (aka eclipse.exe/jmstoolbox.exe) has actually changed.
> 

@jonahgraham Could it be an option to always recompile launcher.exe, and so get a fresh signature on every eclipse release (ie every 3 months)? This "could" mitigate the "problem"
To be honest, as the author of JMSToolBox (JTB), and contributor to a few other RCP apps, this is the first time a user is complaining on this. I do not know if other RCP apps have been confronted to this
On my side I tend to publish a new version of JTB after every eclipse RCP release

**Comment by jonahgraham** [comment]
Thanks for the screenshots. IMHO the error message Beyondtrust has provided you is wrong (or at least misleading). The signature is invalid (ref https://github.com/jmstoolbox/jmstoolbox/issues/173#issuecomment-2204020619), but it is not actually expired (ref https://github.com/jmstoolbox/jmstoolbox/issues/173#issuecomment-2203497356). Of course to users invalid vs expired is effectively the same, it means that the software may not be safe to use.

> I must admit that i do know the difference between JMSToolbox.exe and JMSToolboxc.exe

The difference is ["console"](https://github.com/eclipse-platform/eclipse.platform/blob/0f64336d2bc4c785643f38d429bd36bdbae0337e/docs/FAQ/FAQ_How_do_I_run_Eclipse.md#:~:text=On%20Windows%2C%20the%20eclipsec.exe%20console%20executable%20can%20be%20used%20for%20improved%20command%20line%20behavior.) application vs not. JMSToolboxc.exe is a console version of JMSToolbox.exe, which means it has a stdin/out and generally will open a console window (cmd.exe or similar) when you use it. You shouldn't see any other behaviour differences compared to  JMSToolbox.exe.

**Comment by jonahgraham** [comment]
> @jonahgraham Could it be an option to always recompile launcher.exe, and so get a fresh signature on every eclipse release (ie every 3 months)? This "could" mitigate the "problem"

Since my conclusion is that there is no problem here on Eclipse signing side, but rather on the "normal" way of adding icons that breaks the signature, I don't think this will have any effect. Even in this case eclipse.exe was signed only 3 weeks before release.

FWIW, the Google Chrome I have installed was signed on June 21st with a certificate that expires on July 10, so Eclispe's practices are very similar to other leading software suppliers:

![image](https://github.com/jmstoolbox/jmstoolbox/assets/679236/4fd3f89f-571d-4e44-9298-84a1e792131e)


What would be nice is if there was a way of handling adding icons to eclipse.exe that didn't break signatures so that projects like JTB wouldn't have to setup signing infrastructure just to resign it. Perhaps raise a feature request at eclipse tycho?

>  this is the first time a user is complaining on this

We (the wider Eclipse community) have been seeing such reports more over time as companies take software supply chain security more seriously. Expect more disruption like this in the coming years as various EU and US rules adds more supply chain security requirements (such as SBOM)

**Comment by titou10titou10** [comment]
> What would be nice is if there was a way of handling adding icons to eclipse.exe that didn't break signatures so that projects like JTB wouldn't have to setup signing infrastructure just to resign it. Perhaps raise a feature request at eclipse tycho?

Sure, but the "product" definition and the way to declare app icons is part of eclipse RCP, and tycho is just the "assembler/packager", right? 
So IMHO there are too paths here:
- ask tycho to change the way it processes the product.xml definition file to embed the icon and make the .exe signed at the end
- ask eclipse RCP to change the way icons are defined (or at least processed) and "find a way for an app to configure an icon and have the launcher to process the icon without breaking the signature when an icon definition is present in product.xml"

I know that the second point is quite vague but I'm sure you see what I try to explain and I have no idea if it is even possible, because the icon is embedded in the .exe in windows...

**Comment by netomi** [comment]
just to come back to that topic, I inspected the JMSToolbox.exe binary:

```
tn@proteus:~/Downloads/JMSToolBox$ osslsigncode verify -in JMSToolBox.exe 
Current PE checksum   : 00091CCF
Calculated PE checksum: 00087B9F
Warning: invalid PE checksum

Message digest algorithm  : SHA256
Current message digest    : 3A33BF2895FABCCAE997160969625879F7EB4ADBCA71E23328B421DB51A0A4F3 
Calculated message digest : 5E3C9C7089815087FC64179925EBE6910B72C772ED00DBAA844BCE1EE184FFA1     MISMATCH!!!

Signature verification: failed

Number of verified signatures: 1
Failed
```

which shows that there is a digest mismatch due to the update icon and the binary not being resigned. The expiration date of the signing certificate is not the cause of the problem and just a red herring. In general, expired signing certificates are no problem, only in very specific cases, see: https://stackoverflow.com/questions/329396/what-happens-when-a-code-signing-certificate-expires

Otherwise, many applications would not run on Windows anymore and signing certificates usually have a max validity of 3 years, so you could not run software older than 3 years on Windows.

**Comment by jonahgraham** [comment]
Thanks @netomi for confirmation.

**Comment by jonahgraham** [comment]
> Sure, but the "product" definition and the way to declare app icons is part of eclipse RCP, and tycho is just the "assembler/packager", right?

not really anymore - tycho basically defines it, and feeds back into Eclipse RCP. They are two halves to the same problem/solution. Tycho is a lot more than just the assembler, it is what builds the product, sometimes reusing Eclipse RCP components, but in many places augmenting them significantly. 


> So IMHO there are too paths here:
> 
> - ask tycho to change the way it processes the product.xml definition file to embed the icon and make the .exe signed at the end

The Maven/Tycho workflow already supports this - but you need a signing certificate setup and to do the signing. That is what Eclipse IDE does [here](https://github.com/eclipse-packaging/packages/blob/5099a6ce821107f99240ce1cf037c86e5c477e37/releng/org.eclipse.epp.config/parent/product/pom.xml#L122-L150) using the `eclipse-winsigner-plugin`. The eclipse-winsigner-plugin is designed to work with Eclipse Foundation's signing service which is only accessible for official Eclipse Projects.

There are other signing options, including just running the [signtool](https://learn.microsoft.com/en-us/windows/win32/seccrypto/signtool) directly. This [old blog post entry](https://codeandme.blogspot.com/2013/07/tycho-10-signing-plugins-and-executables.html) probably still works, see step 3 and step 4.

But all these options means having a secure place to store the certificate, and possibly obtaining one as a self-signed certificate is probably insufficient for users like @guidovossen who's employer relies on Beyondtrust.

> - ask eclipse RCP to change the way icons are defined (or at least processed) and "find a way for an app to configure an icon and have the launcher to process the icon without breaking the signature when an icon definition is present in product.xml"

This is a good idea, and perhaps there is a way to do it. But I can't find one in my ~15 minutes of looking around the internet. This [SO](https://stackoverflow.com/a/57588496/2796832) answer vaguely says "externalize the resources, so it's not part of the signature"

> I know that the second point is quite vague but I'm sure you see what I try to explain and I have no idea if it is even possible, because the icon is embedded in the .exe in windows...

**Comment by titou10titou10** [comment]
> But all these options means having a secure place to store the certificate, and possibly obtaining one as a self-signed certificate is probably insufficient for users like @guidovossen who's employer relies on Beyondtrust.

Exact. To be honest, in my mind, signing the product with an officially recognized certificate has been in the air since years, as some Mac users have reported similar difficulties running JMSToolBox as it requires sometime to relax some security rules and I remember a thread about creating a Chocolatey package that requires signed apps...

JMSToolBox is FOSS. I developed it on my spare time "for fun", it's out of scope that I'll invest money in an officially recognized certificate . Maybe a sponsor could make a donation? lol

Anyway, thanks everybody for the investigation around this issue and I'm glad to see that this raised some interesting questions around building and publishing FOSS apps based on eclipse RCP

**Comment by guidovossen** [comment]
for now I'm ok, and indeed it looks like the digital signature is invalid and the 'Digital Certificate Expired' is either a wrong errormessage  , or a too strict validation.   

maybe 3 observations.
1.  JMSToolboxc.exe has a valid cert, and I can run. 
2. JMSToolbox.exe even windows says is invalid. 
3. JMSTooslboxc.exe seems to spawn a process of JMSToolbox.exe and BeyondTrust does not complain... mabye i should log a bug there :) 

![image](https://github.com/jmstoolbox/jmstoolbox/assets/48794999/3349f21a-7a84-415d-9095-9d7dd34babc7)


@titou10titou10  How much is a certificate ?

**Comment by netomi** [comment]
A certificate for digital signature of code can be purchased for example at ssl.com for a rather fair price: https://www.ssl.com/certificates/code-signing/buy/ or even cheaper here: https://shop.certum.eu/open-source-code-signing.html

**Comment by guidovossen** [comment]
@titou10titou10 

I guess the cheapest option is to forfeit on the custom icon as that is breaking the signature..  (i'm assuming that it is right that beyondtrust is actually not complaining about the expired root cert, which makes sense because it is also not complaining about the JMSToolboxc.exe which as the same certificate-hierarchy.)  

But I hate freeloaders so i'm happy to sponsor you the https://shop.certum.eu/open-source-code-signing.html  
if that would allow you to make JMSToolbox.exe trusted.   Just send me a DM.

**Comment by titou10titou10** [comment]
@guidovossen thanks for the offer !
I will evaluate the opportunity to sign the executables and I'll contact you if I go on the process
